### ISSUE TITLE
Dynamite: support ContentString and headers

### ISSUE THREAD

**Issue by Leptopoda** [provokateurin/nextcloud-neon#281]
#194 will remove/degrade the current Support for ContentStrings and headers.
This issue tracks this regression.

TL;DR It'll hopefully come back shortly but it's not as trivial as one might think.

Problem:
`built_value` works on Types instead of fields like `json_serializable`. This means that we can't add the ability to de serialize some fields differently (to built_value they are just strings).

We need to introduce a custom Type for ContentStrings (as done in the mentioned PR). We can't rely on custom logic in it's serializer (`_$ContentStringSerializer`) as this stage doesn't expose the data in a json format. This is especially needed when contentstrings are nested (`ContentString<...<SontentString<...>>>`

<details>
<summary>Why? - Small BuiltValue excursion :)</summary>
<br>
built_value will serialize the data to a custom Iterable<String> based format (`['$', 'TypeWireName', 'FieldWireName', 'serializedContent']`). We can and do apply the `StandardJsonPlugin` wich will take this format once and parse it from/to valid json. This brings improved performance and type support over just serializing to json.

Yes in theory this intermediate format could also then be parsed to xml (FRs exist) [Thinking of moving webdav to it once it exists]
</details>

How can we mitigate this?
We can register our own Plugin that will run before/after the StandardJsonPlugin. This way we can get raw access to the json/intermediate format and manipulate it as needed.

```dart
abstract class ContentString<T> implements Built<ContentString<T>, ContentStringBuilder<T>> {
  factory ContentString([final void Function(ContentStringBuilder<T>)? b]) = _$ContentString<T>;
  const ContentString._();

  /// decoded contentString
  T get content;

  static Serializer<ContentString<Object?>> get serializer => _$contentStringSerializer;
}

class ContentStringPlugin implements SerializerPlugin {
  @override
  Object? afterDeserialize(final Object? object, final FullType specifiedType) => object;

  @override
  Object? afterSerialize(final Object? object, final FullType specifiedType) {
    if (!specifiedType.root.toString().contains('ContentString')) {
      return object;
    }

    if (object is! Map<String, dynamic>) {
      throw ArgumentError('ContentStringPlugin can only be applied to Map<String, dynamic>. '
          'Please ensure the StandardJsonPlugin is applied and run before.');
    }

    final content = object['content'];

    return jsonEncode(content);
  }

  @override
  Object? beforeDeserialize(final Object? object, final FullType specifiedType) {
    if (!specifiedType.root.toString().contains('ContentString')) {
      return object;
    }

    if (object is! String) {
      throw ArgumentError('The serialized ContentString must be of type String. '
          'Please ensure the StandardJsonPlugin is applied and run before.');
    }

    final content = jsonDecode(object);

    return ['content', content];
  }

  @override
  Object? beforeSerialize(final Object? object, final FullType specifiedType) => object;
}
```

Note we throw errors when we detect invalid input indicating that the StandardJsonPlugin hasn't run yet. This can be due to forgetting to register it. Or wrong registration order (it needs to be registered first).

Cool the code is there why isn't it in dynamite?

built_value needs to have a `addBuilderFactory` registered fo every Type, Subtype combination (one for `BuiltList<String>`, one for `BuiltList<int>` ...). The generator will do this for us automatically for every combination it'll find in classes it generates for.
If we have a custom combination that doesn't occur otherwise (because it won't ever happen in a response but is needed for a request) we need to register those ourself. 

This can get ugly and I paused my work thinking about other possibilities before I fully commit to this route.

**Comment by provokateurin** [comment]
This can be closed, right?
